File: src/algebra2_complex_number_equations_medium.py
"""
algebra2 - complex_number_equations (medium)
Generated: 2026-02-22T04:22:14.511603
"""

from problem_utils import *

def generate():
    problem_type = choice(['solve_basic', 'solve_quadratic', 'operations', 'powers'])
    
    if problem_type == 'solve_basic':
        # Solve az + b = c + di where a, b, c, d are real
        # Target difficulty: 1300-1400
        a_coeff = nonzero(-5, 5)
        b_real = randint(-8, 8)
        c_real = randint(-8, 8)
        d_imag = nonzero(-8, 8)
        
        lhs = a_coeff * z + b_real
        rhs = c_real + d_imag * I
        
        solution_val = solve(Eq(lhs, rhs), z)[0]
        
        return problem(
            question=f"Solve for $z$: ${latex(lhs)} = {latex(rhs)}$",
            answer=solution_val,
            difficulty=(1300, 1400),
            topic="algebra2/complex_number_equations",
            solution=steps(
                f"${latex(lhs)} = {latex(rhs)}$",
                f"${latex(a_coeff * z)} = {latex(rhs - b_real)}$",
                f"$z = {latex(solution_val)}$"
            ),
        )
    
    elif problem_type == 'solve_quadratic':
        # Solve z^2 + bz + c = 0 where solutions are complex
        # Target difficulty: 1500-1600
        real_part = randint(-3, 3)
        imag_part = nonzero(-4, 4)
        
        # Roots are real_part ± imag_part*i
        root1 = real_part + imag_part * I
        root2 = real_part - imag_part * I
        
        # Expand (z - root1)(z - root2)
        b_coeff = -2 * real_part
        c_coeff = real_part**2 + imag_part**2
        
        equation = z**2 + b_coeff * z + c_coeff
        solutions = solve(equation, z)
        
        ans = FiniteSet(*solutions)
        
        return problem(
            question=f"Solve for $z$: ${latex(equation)} = 0$",
            answer=ans,
            difficulty=(1500, 1600),
            topic="algebra2/complex_number_equations",
            solution=steps(
                f"Use quadratic formula: $z = \\frac{{-b \\pm \\sqrt{{b^2 - 4ac}}}}{{2a}}$",
                f"$z = \\frac{{-({b_coeff}) \\pm \\sqrt{{({b_coeff})^2 - 4(1)({c_coeff})}}}}{{2}}$",
                f"$z = \\frac{{{-b_coeff} \\pm \\sqrt{{{b_coeff**2 - 4*c_coeff}}}}}{{2}}$",
                f"$z = {latex(root1)}$ or $z = {latex(root2)}$"
            ),
        )
    
    elif problem_type == 'operations':
        # Simplify (a + bi)(c + di) or divide complex numbers
        # Target difficulty: 1300-1500
        if choice([True, False]):
            # Multiplication
            a_val = nonzero(-5, 5)
            b_val = nonzero(-5, 5)
            c_val = nonzero(-5, 5)
            d_val = nonzero(-5, 5)
            
            z1 = a_val + b_val * I
            z2 = c_val + d_val * I
            
            result = expand(z1 * z2)
            
            return problem(
                question=f"Simplify: $({latex(z1)})({latex(z2)})$",
                answer=result,
                difficulty=(1300, 1400),
                topic="algebra2/complex_number_equations",
                solution=steps(
                    f"$({a_val} + {b_val}i)({c_val} + {d_val}i)$",
                    f"$= {a_val*c_val} + {a_val*d_val}i + {b_val*c_val}i + {b_val*d_val}i^2$",
                    f"$= {a_val*c_val} + {a_val*d_val + b_val*c_val}i + {b_val*d_val}(-1)$",
                    f"$= {latex(result)}$"
                ),
            )
        else:
            # Division
            a_val = randint(-6, 6)
            b_val = nonzero(-6, 6)
            c_val = nonzero(-4, 4)
            d_val = nonzero(-4, 4)
            
            numerator = a_val + b_val * I
            denominator = c_val + d_val * I
            
            result = simplify(numerator / denominator)
            
            conjugate = c_val - d_val * I
            
            return problem(
                question=f"Simplify: $\\frac{{{latex(numerator)}}}{{{latex(denominator)}}}$",
                answer=result,
                difficulty=(1400, 1500),
                topic="algebra2/complex_number_equations",
                solution=steps(
                    f"Multiply by conjugate: $\\frac{{{latex(numerator)}}}{{{latex(denominator)}}} \\cdot \\frac{{{latex(conjugate)}}}{{{latex(conjugate)}}}$",
                    f"$= \\frac{{{latex(expand(numerator * conjugate))}}}{{{latex(expand(denominator * conjugate))}}}$",
                    f"$= {latex(result)}$"
                ),
            )
    
    else:  # powers
        # Simplify i^n or solve equations with powers of i
        # Target difficulty: 1300-1400
        if choice([True, False]):
            # Simplify i^n
            n_val = randint(5, 20)
            result = I**n_val
            
            remainder = n_val % 4
            
            return problem(
                question=f"Simplify: $i^{{{n_val}}}$",
                answer=result,
                difficulty=(1300, 1400),
                topic="algebra2/complex_number_equations",
                solution=steps(
                    f"Since $i^4 = 1$, find ${n_val} \\bmod 4 = {remainder}$",
                    f"$i^{{{n_val}}} = i^{{{remainder}}} = {latex(result)}$"
                ),
            )
        else:
            # Solve z^2 = a + bi
            a_val = randint(-8, 8)
            b_val = nonzero(-8, 8)
            
            target = a_val + b_val * I
            solutions = solve(z**2 - target, z)
            ans = FiniteSet(*solutions)
            
            return problem(
                question=f"Solve for $z$: $z^2 = {latex(target)}$",
                answer=ans,
                difficulty=(1500, 1600),
                topic="algebra2/complex_number_equations",
                solution=steps(
                    f"$z^2 = {latex(target)}$",
                    f"$z = \\pm\\sqrt{{{latex(target)}}}$",
                    f"Solutions: $z = {latex(solutions[0])}$ or $z = {latex(solutions[1])}$"
                ),
            )

emit(generate())

File: src/algebra2_complex_number_equations_medium.jl
# algebra2 - complex_number_equations (medium)
# Generated: 2026-03-08T20:14:59.338986

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra2/complex_number_equations")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Quadratic with complex roots: x^2 + bx + c = 0
        # Want discriminant < 0
        real_part = randint(-8, 8)
        imag_part = nonzero(-8, 8)
        
        # Roots are real_part ± imag_part*i
        # Sum of roots = 2*real_part, product = real_part^2 + imag_part^2
        b_coeff = -2 * real_part
        c_coeff = real_part^2 + imag_part^2
        
        expr = x^2 + b_coeff*x + c_coeff
        
        if real_part == 0
            ans = fmt_set(Set(["$(imag_part)i", "$(-imag_part)i"]))
            sol_text = "\\{$(imag_part)i, $(-imag_part)i\\}"
        else
            ans = fmt_set(Set(["$(real_part)+$(imag_part)i", "$(real_part)-$(imag_part)i"]))
            sol_text = "\\{$(real_part)+$(imag_part)i, $(real_part)-$(imag_part)i\\}"
        end
        
        discriminant = b_coeff^2 - 4*c_coeff
        
        problem(
            question = "Solve for \\(x\\): \$$(tex(expr)) = 0\$",
            answer = ans,
            answer_type = "set",
            difficulty = (1200, 1400),
            solution = steps(
                sol("Given", expr ~ 0),
                "Use quadratic formula: \$x = \\frac{-b \\pm \\sqrt{b^2-4ac}}{2a}\$",
                "Discriminant: \$\\Delta = $(b_coeff)^2 - 4(1)($(c_coeff)) = $(discriminant) < 0\$",
                "Solutions: \$$(sol_text)\$"
            ),
            time = 120
        )
        
    elseif problem_type == 2
        # Equation with i: ax + bi = c + di
        a_coeff = nonzero(-10, 10)
        b_coeff = nonzero(-10, 10)
        c_val = randint(-15, 15)
        d_val = randint(-15, 15)
        
        # Solve: a*x_real + b*x_imag*i = c + d*i
        # Real part: a*x_real = c → x_real = c/a
        # Imag part: b*x_imag = d → x_imag = d/b
        
        x_real = c_val // a_coeff
        x_imag = d_val // b_coeff
        
        if x_imag == 0
            ans = "$(x_real)"
        elseif x_real == 0
            ans = "$(x_imag)i"
        else
            ans = "$(x_real)+$(x_imag)i"
        end
        
        problem(
            question = "Solve for the complex number \\(x\\): \$$(a_coeff)x + $(b_coeff)i = $(c_val) + $(d_val)i\$",
            answer = ans,
            difficulty = (1100, 1300),
            solution = steps(
                "Equation: \$$(a_coeff)x + $(b_coeff)i = $(c_val) + $(d_val)i\$",
                "Separate real and imaginary parts: \$$(a_coeff)x = $(c_val)\$ and \$$(b_coeff)i = $(d_val)i\$",
                "Solve for real part: \$x_{real} = \\frac{$(c_val)}{$(a_coeff)} = $(x_real)\$",
                sol("Answer", ans)
            ),
            time = 90
        )
        
    elseif problem_type == 3
        # (x - z1)(x - z2) = 0 where z1, z2 are complex
        r1_real = randint(-8, 8)
        r1_imag = nonzero(-8, 8)
        r2_real = randint(-8, 8)
        r2_imag = nonzero(-8, 8)
        
        # Expand to get coefficients
        # (x - (a+bi))(x - (c+di)) = x^2 - (a+c)x - (bi+di)x + (a+bi)(c+di)
        sum_real = r1_real + r2_real
        sum_imag = r1_imag + r2_imag
        
        prod_real = r1_real * r2_real - r1_imag * r2_imag
        prod_imag = r1_real * r2_imag + r1_imag * r2_real
        
        if sum_imag == 0 && prod_imag == 0
            expr = x^2 - sum_real*x + prod_real
        else
            # Build expression string manually
            b_str = sum_imag == 0 ? "$(sum_real)" : "($(sum_real)+$(sum_imag)i)"
            c_str = prod_imag == 0 ? "$(prod_real)" : "($(prod_real)+$(prod_imag)i)"
            expr_str = "x^2 - $(b_str)x + $(c_str)"
        end
        
        z1_str = r1_imag > 0 ? "$(r1_real)+$(r1_imag)i" : "$(r1_real)$(r1_imag)i"
        z2_str = r2_imag > 0 ? "$(r2_real)+$(r2_imag)i" : "$(r2_real)$(r2_imag)i"
        
        ans = fmt_set(Set([z1_str, z2_str]))
        
        problem(
            question = "Find all complex solutions to \$x^2 - ($(sum_real)+$(sum_imag)i)x + ($(prod_real)+$(prod_imag)i) = 0\$",
            answer = ans,
            answer_type = "set",
            difficulty = (1400, 1600),
            solution = steps(
                "Given quadratic with complex coefficients",
                "Use quadratic formula or factor if possible",
                "The roots are \$$(z1_str)\$ and \$$(z2_str)\$",
                sol("Answer", ans)
            ),
            time = 150
        )
        
    elseif problem_type == 4
        # x^2 = negative number
        k = nonzero(-50, -1)
        sqrt_k = isqrt(-k)
        
        ans = fmt_set(Set(["$(sqrt_k)i", "$(-sqrt_k)i"]))
        
        problem(
            question = "Solve for \\(x\\): \$x^2 = $(k)\$",
            answer = ans,
            answer_type = "set",
            difficulty = (1000, 1200),
            solution = steps(
                sol("Given", x^2 ~ k),
                "Take square root of both sides: \$x = \\pm\\sqrt{$(k)}\$",
                "Since \$\\sqrt{$(k)} = \\sqrt{$(-k)}i = $(sqrt_k)i\$",
                sol("Answer", ans)
            ),
            time = 60
        )
        
    else
        # x^3 = a + bi (one real root, two complex conjugates)
        a_val = nonzero(-20, 20)
        
        # For simplicity, use x^3 = a (real number) to get one real and two complex roots
        # Real root: a^(1/3)
        # Complex roots involve cube roots of unity
        
        real_root = round(Int, sign(a_val) * abs(a_val)^(1/3))
        
        # Check if it's a perfect cube
        if real_root^3 == a_val
            # omega = e^(2πi/3) = -1/2 + sqrt(3)/2 i
            # roots: real_root, real_root*omega, real_root*omega^2
            
            ans = "$(real_root)"
            
            problem(
                question = "Find the real solution to \$x^3 = $(a_val)\$",
                answer = ans,
                difficulty = (1200, 1400),
                solution = steps(
                    sol("Given", x^3 ~ a_val),
                    "Take cube root: \$x = \\sqrt[3]{$(a_val)}\$",
                    "The real solution is \$x = $(real_root)\$",
                    sol("Answer", ans)
                ),
                time = 90
            )
        else
            # Just do x^2 + k = 0 instead
            k = randint(5, 40)
            sqrt_k = isqrt(k)
            
            ans = fmt_set(Set(["$(sqrt_k)i", "$(-sqrt_k)i"]))
            
            problem(
                question = "Solve for \\(x\\): \$x^2 + $(k) = 0\$",
                answer = ans,
                answer_type = "set",
                difficulty = (1100, 1300),
                solution = steps(
                    sol("Given", x^2 + k ~ 0),
                    "Rearrange: \$x^2 = $(-k)\$",
                    "Take square root: \$x = \\pm $(sqrt_k)i\$",
                    sol("Answer", ans)
                ),
                time = 75
            )
        end
    end
end